File: practice/merge_sort.py
def merge_sort(arr):

    if len(arr) < 2:
        return arr

    mid = len(arr) // 2

    left = merge_sort(arr[0:mid])
    right = merge_sort(arr[mid : len(arr)])

    sorted_arr = []

    li = ri = 0
    while li < mid and ri < len(arr) - mid:
        if left[li] > right[ri]:
            sorted_arr.append(left[li])
            li += 1
        else:
            sorted_arr.append(right[ri])
            ri += 1

    if ri < len(arr) - mid:
        sorted_arr.extend(right[ri:])
    if li < mid:
        sorted_arr.extend(left[li:])

    return sorted_arr


if __name__ == "__main__":
    arr = [3, 4, 1, 2, 6]
    print(merge_sort(arr))

File: practice/merge_sort.cpp
#include <iostream>
using namespace std;

void merge(int arr[], int l, int m, int r)
{
    int ln = m - l + 1;
    int rn = r - m;

    int left[ln], right[rn];

    for (int i = 0; i < ln; i++)
        left[i] = arr[l + i];

    for (int i = 0; i < rn; i++)
        right[i] = arr[m + 1 + i];

    int i = l, li = 0, ri = 0;
    while (li < ln && ri < rn)
    {
        bool rightGreat = left[li] < right[ri];
        arr[i++] = rightGreat ? right[ri++] : left[li++];
    }

    while (li < ln)
        arr[i++] = left[li++];
    while (ri < rn)
        arr[i++] = right[ri++];
}

void merge_sort(int arr[], int l, int r)
{
    if (r <= l)
        return;

    int m = (r + l) / 2;

    merge_sort(arr, l, m);
    merge_sort(arr, m + 1, r);
    merge(arr, l, m, r);
}
int main()
{
    int arr[] = {3, 4, 1, 2, 6, 5};
    merge_sort(arr, 0, 5);
    for (int i = 0; i < 6; i++)
    {
        cout << arr[i] << " ";
    }
    return 0;
}